### PR TITLE
[ADD] spreadsheet_oca & spreadsheet_dashboard_oca

### PR DESCRIPTION
**Pull request by etobella**
This PR contains two modules:
- spreadsheet_oca: allows to edit spreadsheets using a new model
- spreadsheet_dashboard_oca: allows to edit spreadsheet dashboards using the code implemented on the first module

TODO:
- [x] Add a button on pivot table for exporting
- [x] Save Spreadsheet on exit (on first option is only saved when we select it)
- [x] Allow multiple people to edit at the same time
- [x] Improve security and allow to edit / share

https://user-images.githubusercontent.com/28590170/210117043-65740b18-a368-41b6-bc03-1cde568af555.mp4

@pedrobaeza

### CODE REVIEW

**Review by legalsylvain — COMMENTED** [review]
Well. 
First, thanks a lot @etobella for this nice OCA christmas gift ! 

just tested very quickly on runboat : 
- going in spreadsheet spreadsheet, the "create" button doesn't work. Do you face the same issue ? 
- when I go in Dashboard > configuration > Sale (for exemple) > and I edit the sale dashboard : 
-> The editor is well opened : OK
-> When I edit in multi browser with demo and admin, I don't see direct changes : KO. Edit : did'nt see your roadmap point.
-> when I try to save the changes (I click on save), It doesn't work : KO

**Comment by etobella** [comment]
Saving on dashboards now works. About the create, I forgot about it. At the begining I tried to do it with a Form JS Class, but at the end I had to find another way. I need to modify tree view. I will do ASAP :wink:

**Comment by legalsylvain** [comment]
Thanks for your answer ! Please ping me when you want review. 
Happy new years !

**Comment by etobella** [comment]
Tree modified. Creation is allowed properly :tada:

**Comment by etobella** [comment]
Just one thing. Should we use `dms` functionality in order to avoid to redo folders and permissions? 

Next step would be to improve views on DMS. But it might be achieved :smile:

**Comment by pedrobaeza** [comment]
Yeah, I was thinking in a `spreadsheet_dms_oca` module. For this one, I think we should set a basic security without folders, having groups per spreadsheet. Forcing to use DMS for a basic security seems too much for me.

File: spreadsheet_oca/views/spreadsheet_spreadsheet.xml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    <record model="ir.actions.act_window" id="spreadsheet_spreadsheet_act_window">
+        <field name="name">Spreadsheet Spreadsheet</field> <!-- TODO -->
+        <field name="res_model">spreadsheet.spreadsheet</field>
+        <field name="view_mode">tree</field>
+        <field name="domain">[]</field>
+        <field name="context">{}</field>
+    </record>
+
+    <record model="ir.ui.menu" id="spreadsheet_spreadsheet_menu">
+        <field name="name">Spreadsheet Spreadsheet</field>
```
Comment (legalsylvain) [inline spreadsheet_oca/views/spreadsheet_spreadsheet.xml]:
```suggestion
        <field name="name">Spreadsheets</field>
```

File: spreadsheet_oca/views/spreadsheet_spreadsheet.xml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        <field name="model">spreadsheet.spreadsheet</field>
+        <field name="arch" type="xml">
+            <tree js_class="spreadsheet_list">
+                <field name="name" />
+            </tree>
+        </field>
+    </record>
+
+    <record model="ir.actions.act_window" id="spreadsheet_spreadsheet_act_window">
+        <field name="name">Spreadsheet Spreadsheet</field> <!-- TODO -->
```
Comment (legalsylvain) [inline spreadsheet_oca/views/spreadsheet_spreadsheet.xml]:
```suggestion
        <field name="name">Spreadsheets</field>
```

**Review by legalsylvain — COMMENTED** [review]
Hi @etobella. Thanks for your work ! Saving works now.

remarks : 
- Nice to fix : minor remarks inline, regarding the name of "spreadsheets" menu.
- Question : I don't find a way to add a new spreadsheet to a spreadsheet group. Is it possible ? forecast ? 
- UI remarks : I see the current UI in spreadsheet spreadsheet list very limited. It is only possible to create and edit via ``[o-spreadsheet](https://github.com/odoo/o-spreadsheet)`` the spreadsheet. but it's not possible to open the form. As a result, if OCA or custom module want to add some field, it will not be possible to edit them. Why not having a classic "tree" and "form" views here, with a simple button "Open" or "Edit" when we want to edit the content ? 

thanks !

**Comment by etobella** [comment]
Hi @legalsylvain 

Just a few things:
- multi-editing should be working now
- Menu names, fixed
- About your questions, I am not sure what are you asking. Do you mean to access another spreadsheet? Might be possible, but don't know how. Using the same logic from ODOO.BALANCE. It might be possible.
- About opening the form, I was thinking on something different. IMO the main option should be to use the spreadsheet and adding a button to access the form on the spreadsheet view (near the name using a +config or something like that)

Multi edit video :smile: 

https://user-images.githubusercontent.com/28590170/210182863-3a2f98fa-cc87-449b-8dbb-434b5976fe52.mp4

**Comment by legalsylvain** [comment]
> Menu names, fixed

The last instance on runboat have the same "spreadsheet spreasheet" title.

> About your questions, I am not sure what are you asking. Do you mean to access another spreadsheet? Might be possible, but don't know how. Using the same logic from ODOO.BALANCE. It might be possible.

I mean : Go to "Dashboard" > "Configuration" > Click on a group (like "Sale") and here, have the possibility to "Add a line".

image

Does it makes sense ? 

> IMO the main option should be to use the spreadsheet and adding a button to access the form on the spreadsheet view

As you wish. I guess, that having a "classic" tree / form view structure and just add a button "Open" in the tree view (and in the form view) could be the more simple to implement and the UI the most "awaited". (I mean, all the rest of odoo works the same, when you click on a line of a tree view, the form view is opened). 
Anyway, implementing a form view is a must have, because otherwise, no "duplicate" function is available.
Not a blocking point, thought.

> Multi edit

1) Tested in spreadsheet spreadsheet entry. Work like a charm ! Nice ! Wow effect !

2) Regression : go to Dashboard / configuration / open a sheet : You have an error : 

``AttributeError: type object 'spreadsheet.dashboard' has no attribute 'send_spreadsheet_message'``

I guess you should implement ``send_spreadsheet_message`` in this model too.

Side questions : 

- I see that it's possible to have spreadsheets with no name. (when creating a new one, by default). Maybe having a default "Untitled" or "No name" could be more appropriate.

**Comment by etobella** [comment]
> I mean : Go to "Dashboard" > "Configuration" > Click on a group (like "Sale") and here, have the possibility to "Add a line".
> Does it makes sense ?
Yes! Fixed :smile: 

> > IMO the main option should be to use the spreadsheet and adding a button to access the form on the spreadsheet view
> 
> As you wish. I guess, that having a "classic" tree / form view structure and just add a button "Open" in the tree view (and in the form view) could be the more simple to implement and the UI the most "awaited". (I mean, all the rest of odoo works the same, when you click on a line of a tree view, the form view is opened). Anyway, implementing a form view is a must have, because otherwise, no "duplicate" function is available. Not a blocking point, thought.

After a second thought, I think you are right. Changes applied :smile: 
 
> > Multi edit
> 
> 1. Tested in spreadsheet spreadsheet entry. Work like a charm ! Nice ! Wow effect !
> 2. Regression : go to Dashboard / configuration / open a sheet : You have an error :
> 
> `AttributeError: type object 'spreadsheet.dashboard' has no attribute 'send_spreadsheet_message'`
> 
> I guess you should implement `send_spreadsheet_message` in this model too.

I passed the logic to an abstract in order to avoid the same error on the future :wink: 
 
> Side questions :
> 
> * I see that it's possible to have spreadsheets with no name. (when creating a new one, by default). Maybe having a default "Untitled" or "No name" could be more appropriate.

:thinking:  Maybe we need to make it required

**Comment by legalsylvain** [comment]
Thanks for all the changes. Nice refactoring with the abstract btw !

**Review by chienandalu — APPROVED** [review]
Awesome work @etobella ! Thanks!

Some features that won't be hard to implement in a near future:

- Link lists, and pivots to new or existing spreadsheets. Checking in the tests the expected json it won't take too much to create a client.action that prepares the proper format (ref: https://github.com/OCA/OCB/blob/16.0/addons/spreadsheet/static/tests/global_filters/global_filters_model_test.js#L942-L953)
- Exporting to be able to download the json and use it elsewhere.
- Importing a json to retrieve exported spreadsheets.
- A module linking spreadsheets and dashboards  to use the data from the former ones as the dashboard source.

**Comment by etobella** [comment]
We can now export from Pivot Tables :tada: 

https://user-images.githubusercontent.com/28590170/210276196-e6ec4978-425d-4bc5-8a86-c5f9edaeb2f4.mp4

**Comment by chienandalu** [comment]
Bravo!

File: spreadsheet_oca/views/spreadsheet_spreadsheet.xml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    <record model="ir.actions.act_window" id="spreadsheet_spreadsheet_act_window">
+        <field name="name">Spreadsheet</field>
+        <field name="res_model">spreadsheet.spreadsheet</field>
+        <field name="view_mode">tree,form</field>
+        <field name="domain">[]</field>
+        <field name="context">{}</field>
+    </record>
+
+    <record model="ir.ui.menu" id="spreadsheet_spreadsheet_menu">
+        <field name="name">Spreadsheet</field>
```
Comment (pedrobaeza) [inline spreadsheet_oca/views/spreadsheet_spreadsheet.xml]:
```suggestion
        <field name="name">Spreadsheets</field>
```

File: spreadsheet_oca/security/security.xml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record model="ir.rule" id="spreadsheet_owner_rule">
+        <field name="name">Spreadsheet Owner</field>
+        <field name="model_id" ref="model_spreadsheet_spreadsheet" />
+        <field name="groups" eval="[(4, ref('base.group_user'))]" />
+        <field name="domain_force">[('owner_id','=', user.id)]</field>
+    </record>
+    <record model="ir.rule" id="spreadsheet_contributor_rule">
+        <field name="name">Spreadsheet Owner</field>
```
Comment (pedrobaeza) [inline spreadsheet_oca/security/security.xml]:
```suggestion
        <field name="name">Spreadsheet Contributor</field>
```

File: spreadsheet_oca/models/spreadsheet_spreadsheet.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    data = fields.Binary()
+    spreadsheet_raw = fields.Serialized(
+        compute="_compute_spreadsheet_raw", inverse="_inverse_spreadsheet_raw"
+    )
+    owner_id = fields.Many2one(
+        "res.users", required=True, default=lambda r: r.env.user.id
+    )
+    contributor_ids = fields.Many2many(
+        "res.users",
```
Comment (pedrobaeza) [inline spreadsheet_oca/models/spreadsheet_spreadsheet.py:24]:
```suggestion
        "res.users",
        string="Contributors",
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Thank you very much for this.

Several things to finish the modules:

- The dashboard edition doesn't work:

Peek 03-01-2023 16-28

- On this matter, I prefer that the dashboards with XML-ID can't be overwritten, and only allows to save a copy.

- I also misses the possibility to save a spreadsheet as a dashboard.

- About security, I would create 2 groups:

Spreadsheets / User
Spreadsheets / Manager

You may want that your users don't access spreadsheets at all, so not putting any of them, the main menu is not shown, as well as the button on the pivot views. If you put user, then you will be able to see those spreadsheets that you are owner, contributor or reader, and edit the first 2. Being manager, you see and edit all.

- From the pivot view, is there any possibility of adding it to an existing spreadsheet? It would be a very nice feature.

- Spreadsheets can be saved with no name, so please put it as required.

Maybe it's better to split the pull request in 2 parts and let this one for the base module and handle the dashboard thing in another.

Thanks again.

**Comment by pedrobaeza** [comment]
Please add also this icon, and put it on the menu (it can directly be the SVG I think).

icon

**Comment by etobella** [comment]
Hi @pedrobaeza Your inline comments have been attended

About your main comment :smile: 

> * The dashboard edition doesn't work:

True, should be fixed now.

> * On this matter, I prefer that the dashboards with XML-ID can't be overwritten, and only allows to save a copy.

Well, I understand your concern, but I am not sure if it is the right approach :thinking: Also, the system checks who can edit it, and it is only editable by the right users, so only administrator users are allowed to edit spreadsheet dashboards.

> * I also misses the possibility to save a spreadsheet as a dashboard.

The idea is to create it directly there on the dashboard configuration.

> * About security, I would create 2 groups:
> 
> Spreadsheets / User Spreadsheets / Manager
> You may want that your users don't access spreadsheets at all, so not putting any of them, the main menu is not shown, as well as the button on the pivot views. If you put user, then you will be able to see those spreadsheets that you are owner, contributor or reader, and edit the first 2. Being manager, you see and edit all.
> 
It makes sense, I will do it ASAP

> * From the pivot view, is there any possibility of adding it to an existing spreadsheet? It would be a very nice feature.

It is already done :smile: 

> * Spreadsheets can be saved with no name, so please put it as required.

Yes, ASAP :wink: 

> Maybe it's better to split the pull request in 2 parts and let this one for the base module and handle the dashboard thing in another.

I will close this one when it is ready in order to keep the history in this PR. then I will create two new ones with only one commit

Also, thanks for the icon, it is already here :smile:

**Comment by pedrobaeza** [comment]
OK, tested. The Edit button still appears when you don't have any Spreadsheet group, but then it appears an error when you click it. Maybe it's better to hide such button. About having that edit button on standard dashboards, it seems it's still usable for spreadsheet / user.

You can now proceed to clean the commit history and let only the main module here.

**Comment by etobella** [comment]
Done!

**Pull request closed without merging**